lib/SILGen/SILGenPoly.cpp

```diff
@@ -509,7 +509,8 @@ ManagedValue Transform::transform(ManagedValue v,
   bool inputIsOptional = (bool) inputObjectType;
 
   CanType outputObjectType = outputSubstType.getOptionalObjectType();
-  bool outputIsOptional = (bool) outputObjectType;
+  bool outputIsOptional =
+      static_cast<bool>(loweredResultTy.getASTType().getOptionalObjectType());
```

Comment:
This seems like the wrong place to do this. If there's an implicit level of optionality introduced around these APIs, we should track back to a place that matches that decision and have that place pass down an optional substituted type.